src/tools/remote-test-server/src/main.rs

```diff
@@ -336,23 +341,38 @@ fn set_permissions(path: &Path) {
 #[cfg(windows)]
 fn set_permissions(_path: &Path) {}
 
-fn my_copy(src: &mut dyn Read, which: u8, dst: &Mutex<dyn Write>) {
+fn read2(out_pipe: &mut dyn Read, err_pipe: &mut dyn Read, dst: &Mutex<dyn Write>) {
     let mut b = [0; 1024];
     loop {
-        let n = t!(src.read(&mut b));
-        let mut dst = dst.lock().unwrap();
-        t!(dst.write_all(&[
-            which,
-            (n >> 24) as u8,
-            (n >> 16) as u8,
-            (n >> 8) as u8,
-            (n >> 0) as u8,
-        ]));
-        if n > 0 {
-            t!(dst.write_all(&b[..n]));
-        } else {
-            break;
+        match (read_step(out_pipe, 0, dst, &mut b), read_step(err_pipe, 1, dst, &mut b)) {
```

Comment:
This isn't enough. You're still blocking on the read from the stdout pipe here, which means that if the process only writes to stderr it will hang. The read2 file I linked handles that by setting the file descriptors to non-blocking, but you're not doing that here.

Reply:
Well, the read2 implementation only does that for unix and windows. On all the other targets, it uses [this](https://github.com/rust-lang/rust/blob/a161a7b654083a881b22908a475988bcc3221a79/src/tools/compiletest/src/read2.rs#L127)

Reply:
Is the remote test server here running in the qemu image (i.e., maybe not on unix or windows)?

In any case, that version of read2 has the same problem as here, but we're rarely running cargo or compiletest outside of cfg(unix) or cfg(windows), maybe even never.